File: frontend/lib/src/features/shipment/domain/shipment.dart
/// Shipment status enum representing the current state of a shipment
enum ShipmentStatus {
  processing,
  pickedUp,
  inTransit,
  customs,
  customsCleared,
  arrivedAtHub,
  outForDelivery,
  delivered,
  exception,
  returned,
  cancelled;

  /// Display name for UI presentation
  String get displayName {
    switch (this) {
      case ShipmentStatus.processing:
        return 'Processing';
      case ShipmentStatus.pickedUp:
        return 'Picked Up';
      case ShipmentStatus.inTransit:
        return 'In Transit';
      case ShipmentStatus.customs:
        return 'At Customs';
      case ShipmentStatus.customsCleared:
        return 'Customs Cleared';
      case ShipmentStatus.arrivedAtHub:
        return 'Arrived at Hub';
      case ShipmentStatus.outForDelivery:
        return 'Out for Delivery';
      case ShipmentStatus.delivered:
        return 'Delivered';
      case ShipmentStatus.exception:
        return 'Exception';
      case ShipmentStatus.returned:
        return 'Returned';
      case ShipmentStatus.cancelled:
        return 'Cancelled';
    }
  }

  /// Convert status name to enum value (for MongoDB deserialization)
  static ShipmentStatus fromString(String status) {
    switch (status.toLowerCase().replaceAll(' ', '').replaceAll('_', '')) {
      case 'processing':
        return ShipmentStatus.processing;
      case 'pickedup':
        return ShipmentStatus.pickedUp;
      case 'intransit':
        return ShipmentStatus.inTransit;
      case 'customs':
        return ShipmentStatus.customs;
      case 'customscleared':
        return ShipmentStatus.customsCleared;
      case 'arrivedathub':
        return ShipmentStatus.arrivedAtHub;
      case 'outfordelivery':
        return ShipmentStatus.outForDelivery;
      case 'delivered':
        return ShipmentStatus.delivered;
      case 'exception':
        return ShipmentStatus.exception;
      case 'returned':
        return ShipmentStatus.returned;
      case 'cancelled':
        return ShipmentStatus.cancelled;
      // Legacy status mappings
      case 'pending':
        return ShipmentStatus.processing;
      default:
        return ShipmentStatus.processing;
    }
  }
}

/// Represents a shipment entity matching the MongoDB schema
class Shipment {
  /// Unique identifier (maps to MongoDB _id)
  final String id;

  /// Unique tracking number for the shipment
  final String trackingNumber;

  /// Origin location/address
  final String origin;

  /// Destination location/address
  final String destination;

  /// Current status of the shipment
  final String status;

  /// Estimated delivery date and time
  final DateTime estimatedDelivery;

  /// Created date and time
  final DateTime createdAt;

  /// List of associated package IDs
  final List<String> packageIds;

  /// Customer Name (from clientId)
  final String customerName;

  /// Shipping Cost
  final double cost;

  /// Package Count
  final int packageCount;

  /// Total Weight
  final double totalWeight;

  /// Current Latitude
  final double? currentLat;

  /// Current Longitude
  final double? currentLng;

  /// Destination Latitude
  final double? destLat;

  /// Destination Longitude
  final double? destLng;

  /// Shipment Mode (Air, Sea, etc.)
  final String mode;

  const Shipment({
    required this.id,
    required this.trackingNumber,
    required this.origin,
    required this.destination,
    required this.status,
    required this.estimatedDelivery,
    required this.packageIds,
    this.customerName = 'Unknown',
    this.cost = 0.0,
    this.packageCount = 1,
    this.totalWeight = 0.0,
    required this.createdAt,
    this.currentLat,
    this.currentLng,
    this.destLat,
    this.destLng,
    this.mode = 'Air',
  });

  /// Creates a copy of this [Shipment] with the given fields replaced
  Shipment copyWith({
    String? id,
    String? trackingNumber,
    String? origin,
    String? destination,
    String? status,
    DateTime? estimatedDelivery,
    List<String>? packageIds,
    String? customerName,
    double? cost,
    int? packageCount,
    double? totalWeight,
    DateTime? createdAt,
    double? currentLat,
    double? currentLng,
    double? destLat,
    double? destLng,
    String? mode,
  }) {
    return Shipment(
      id: id ?? this.id,
      trackingNumber: trackingNumber ?? this.trackingNumber,
      origin: origin ?? this.origin,
      destination: destination ?? this.destination,
      status: status ?? this.status,
      estimatedDelivery: estimatedDelivery ?? this.estimatedDelivery,
      packageIds: packageIds ?? this.packageIds,
      customerName: customerName ?? this.customerName,
      cost: cost ?? this.cost,
      packageCount: packageCount ?? this.packageCount,
      totalWeight: totalWeight ?? this.totalWeight,
      createdAt: createdAt ?? this.createdAt,
      currentLat: currentLat ?? this.currentLat,
      currentLng: currentLng ?? this.currentLng,
      destLat: destLat ?? this.destLat,
      destLng: destLng ?? this.destLng,
      mode: mode ?? this.mode,
    );
  }

  /// Creates a [Shipment] from a JSON map (MongoDB document)
  factory Shipment.fromJson(Map<String, dynamic> json) {
    // Handle origin
    String originStr;
    if (json['origin'] is String) {
      originStr = json['origin'] as String;
    } else if (json['origin'] is Map) {
      final originMap = json['origin'] as Map<String, dynamic>;
      originStr = '${originMap['city'] ?? ''}, ${originMap['country'] ?? ''}';
    } else {
      originStr = 'Unknown';
    }

    // Handle destination
    String destinationStr;
    double? dLat;
    double? dLng;
    if (json['destination'] is String) {
      destinationStr = json['destination'] as String;
    } else if (json['destination'] is Map) {
      final destMap = json['destination'] as Map<String, dynamic>;
      destinationStr = '${destMap['city'] ?? ''}, ${destMap['country'] ?? ''}';

      // Parse coordinates if available
      if (destMap['coordinates'] is Map) {
        dLat = (destMap['coordinates']['latitude'] as num?)?.toDouble();
        dLng = (destMap['coordinates']['longitude'] as num?)?.toDouble();
      }
    } else {
      destinationStr = 'Unknown';
    }

    // Handle estimatedDelivery
    DateTime estimatedDelivery;
    if (json['estimatedDelivery'] != null) {
      estimatedDelivery = DateTime.parse(json['estimatedDelivery'] as String);
    } else {
      estimatedDelivery = DateTime.now().add(const Duration(days: 7));
    }

    // Handle createdAt
    DateTime createdAt;
    if (json['createdAt'] != null) {
      createdAt = DateTime.parse(json['createdAt'] as String);
    } else {
      createdAt = DateTime.now();
    }

    // Handle customerName
    String customerName = 'Unknown';
    if (json['clientId'] is Map) {
      customerName = json['clientId']['fullName'] ?? 'Unknown';
    }

    // Handle current location
    double? lat;
    double? lng;
    if (json['currentLocation'] is Map) {
      final loc = json['currentLocation'] as Map<String, dynamic>;
      if (loc['coordinates'] is Map) {
        lat = (loc['coordinates']['latitude'] as num?)?.toDouble();
        lng = (loc['coordinates']['longitude'] as num?)?.toDouble();
      }
    }

    return Shipment(
      id: (json['_id'] ?? json['id'] ?? '') as String,
      trackingNumber: (json['trackingNumber'] ?? '') as String,
      origin: originStr,
      destination: destinationStr,
      status: (json['status'] ?? 'Processing') as String,
      estimatedDelivery: estimatedDelivery,
      packageIds:
          (json['packageIds'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      customerName: customerName,
      cost: (json['shippingCost'] as num?)?.toDouble() ?? 0.0,
      packageCount: (json['packageCount'] as num?)?.toInt() ?? 1,
      totalWeight: (json['totalWeight'] as num?)?.toDouble() ?? 0.0,
      createdAt: createdAt,
      currentLat: lat,
      currentLng: lng,
      destLat: dLat,
      destLng: dLng,
      mode: (json['mode'] ?? 'Air') as String,
    );
  }

  /// Converts this [Shipment] to a JSON map for API requests
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'trackingNumber': trackingNumber,
      'origin': origin,
      'destination': destination,
      'status': status,
      'estimatedDelivery': estimatedDelivery.toIso8601String(),
      'packageIds': packageIds,
      'customerName': customerName,
      'shippingCost': cost,
      'packageCount': packageCount,
      'totalWeight': totalWeight,
      'createdAt': createdAt.toIso8601String(),
      'mode': mode,
    };
  }

  /// Returns the [ShipmentStatus] enum from the status string
  ShipmentStatus get statusEnum => ShipmentStatus.fromString(status);

  @override
  String toString() {
    return 'Shipment(id: $id, trackingNumber: $trackingNumber, status: $status)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Shipment && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}

File: backend/models/Shipment.js
const mongoose = require('mongoose');

// ============================================
// Location Sub-Schema (Embedded Document)
// ============================================
const locationSchema = new mongoose.Schema({
    city: {
        type: String,
        required: [true, 'City is required'],
        trim: true,
    },
    country: {
        type: String,
        required: [true, 'Country is required'],
        trim: true,
    },
    code: {
        type: String,
        trim: true,
        uppercase: true,
        default: '',
    },
    address: {
        type: String,
        trim: true,
        default: '',
    },
    coordinates: {
        latitude: { type: Number },
        longitude: { type: Number },
    },
}, { _id: false });

// ============================================
// Timeline Event Sub-Schema (Embedded Document)
// ============================================
const timelineEventSchema = new mongoose.Schema({
    status: {
        type: String,
        required: [true, 'Event status is required'],
        trim: true,
    },
    location: {
        type: String,
        trim: true,
        default: '',
    },
    timestamp: {
        type: Date,
        required: [true, 'Event timestamp is required'],
        default: Date.now,
    },
    description: {
        type: String,
        trim: true,
        maxlength: [500, 'Description cannot exceed 500 characters'],
        default: '',
    },
    isNotified: {
        type: Boolean,
        default: false,
    },
}, { _id: true }); // Keep _id for individual event tracking

// ============================================
// Shipment Schema
// ============================================
const shipmentSchema = new mongoose.Schema({
    // --- Relations ---
    requestId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'ShipmentRequest',
        index: true,
    },
    quotationId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Quotation',
        index: true,
    },
    clientId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: [true, 'Client ID is required'],
        index: true,
    },
    managerId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        index: true,
    },

    // --- Tracking Information ---
    trackingNumber: {
        type: String,
        required: [true, 'Tracking number is required'],
        unique: true,
        uppercase: true,
        trim: true,
        index: true,
    },
    carrierTrackingNumber: {
        type: String,
        trim: true,
        default: '',
    },
    carrier: {
        type: String,
        trim: true,
        default: '',
    },

    // --- Shipment Details ---
    mode: {
        type: String,
        enum: {
            values: ['Air', 'Sea', 'Land', 'Rail', 'Multimodal'],
            message: 'Invalid shipping mode',
        },
        default: 'Air',
    },
    serviceType: {
        type: String,
        enum: ['Standard', 'Express', 'Economy', 'Priority'],
        default: 'Standard',
    },

    // --- Route Information ---
    origin: {
        type: locationSchema,
        required: [true, 'Origin is required'],
    },
    destination: {
        type: locationSchema,
        required: [true, 'Destination is required'],
    },
    currentLocation: {
        type: locationSchema,
    },

    // --- Package Details ---
    packageCount: {
        type: Number,
        default: 1,
        min: [1, 'Package count must be at least 1'],
    },
    totalWeight: {
        type: Number,
        min: [0, 'Weight cannot be negative'],
        default: 0,
    },
    weightUnit: {
        type: String,
        enum: ['kg', 'lbs'],
        default: 'kg',
    },
    totalCbm: {
        type: Number,
        min: [0, 'CBM cannot be negative'],
        default: 0,
    },
    itemDescription: {
        type: String,
        trim: true,
        default: '',
    },

    // --- Status Management ---
    status: {
        type: String,
        enum: {
            values: [
                'Processing',
                'Picked Up',
                'In Transit',
                'Customs',
                'Customs Cleared',
                'Arrived at Hub',
                'Out for Delivery',
                'Delivered',
                'Exception',
                'Returned',
                'Cancelled'
            ],
            message: 'Invalid shipment status',
        },
        default: 'Processing',
        index: true,
    },
    statusDetails: {
        type: String,
        trim: true,
        default: '',
    },

    // --- Timeline/History ---
    timeline: {
        type: [timelineEventSchema],
        default: [],
    },

    // --- Dates ---
    estimatedDelivery: {
        type: Date,
    },
    actualDelivery: {
        type: Date,
    },
    pickupDate: {
        type: Date,
    },
    departureDate: {
        type: Date,
    },
    arrivalDate: {
        type: Date,
    },

    // --- Customs Information ---
    customsStatus: {
        type: String,
        enum: ['Not Required', 'Pending', 'In Progress', 'Cleared', 'Hold', 'Rejected'],
        default: 'Not Required',
    },
    customsDeclarationNumber: {
        type: String,
        trim: true,
        default: '',
    },
    hsCode: {
        type: String,
        trim: true,
        default: '',
    },

    // --- Financial ---
    declaredValue: {
        type: Number,
        min: [0, 'Declared value cannot be negative'],
        default: 0,
    },
    declaredValueCurrency: {
        type: String,
        default: 'USD',
    },
    shippingCost: {
        type: Number,
        min: [0, 'Shipping cost cannot be negative'],
        default: 0,
    },
    isPaid: {
        type: Boolean,
        default: false,
    },

    // --- Delivery Information ---
    signatureRequired: {
        type: Boolean,
        default: false,
    },
    deliveryInstructions: {
        type: String,
        trim: true,
        maxlength: [500, 'Delivery instructions cannot exceed 500 characters'],
        default: '',
    },
    recipientName: {
        type: String,
        trim: true,
        default: '',
    },
    recipientSignatureUrl: {
        type: String,
        trim: true,
        default: '',
    },

    // --- Documents ---
    documents: [{
        name: { type: String, trim: true },
        type: { type: String, enum: ['invoice', 'packing_list', 'customs', 'pod', 'other'] },
        url: { type: String, trim: true },
        uploadedAt: { type: Date, default: Date.now },
    }],

    // --- Flags ---
    isInsured: {
        type: Boolean,
        default: false,
    },
    isFragile: {
        type: Boolean,
        default: false,
    },
    requiresRefrigeration: {
        type: Boolean,
        default: false,
    },
    isHazardous: {
        type: Boolean,
        default: false,
    },

    // --- Notes ---
    internalNotes: {
        type: String,
        trim: true,
        maxlength: [1000, 'Internal notes cannot exceed 1000 characters'],
        default: '',
    },

}, {
    timestamps: true,
    toJSON: {
        virtuals: true,
        versionKey: false,
        transform: function (doc, ret) {
            ret.id = ret._id;
            delete ret._id;
            delete ret.internalNotes; // Hide from client responses
        }
    },
    toObject: {
        virtuals: true,
        versionKey: false,
        transform: function (doc, ret) {
            ret.id = ret._id;
            delete ret._id;
        }
    }
});

// ============================================
// Indexes for Performance
// ============================================
shipmentSchema.index({ trackingNumber: 1 }, { unique: true });
shipmentSchema.index({ clientId: 1, status: 1 });
shipmentSchema.index({ status: 1, createdAt: -1 });
shipmentSchema.index({ estimatedDelivery: 1 });
shipmentSchema.index({ 'origin.country': 1, 'destination.country': 1 });
shipmentSchema.index({ carrier: 1, status: 1 });

// ============================================
// Virtual Fields
// ============================================

/**
 * Check if shipment is active
 */
shipmentSchema.virtual('isActive').get(function () {
    return !['Delivered', 'Returned', 'Cancelled'].includes(this.status);
});

/**
 * Check if shipment is delivered
 */
shipmentSchema.virtual('isDelivered').get(function () {
    return this.status === 'Delivered';
});

/**
 * Get days in transit
 */
shipmentSchema.virtual('daysInTransit').get(function () {
    const startDate = this.pickupDate || this.createdAt;
    const endDate = this.actualDelivery || new Date();
    const diffTime = Math.abs(endDate - startDate);
    return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
});

/**
 * Get latest status from timeline
 */
shipmentSchema.virtual('latestEvent').get(function () {
    if (!this.timeline || this.timeline.length === 0) {
        return null;
    }
    return this.timeline[this.timeline.length - 1];
});

/**
 * Route summary string
 */
shipmentSchema.virtual('routeSummary').get(function () {
    const origin = this.origin?.code || this.origin?.city || 'Unknown';
    const destination = this.destination?.code || this.destination?.city || 'Unknown';
    return `${origin} → ${destination}`;
});

// ============================================
// Pre-save Middleware
// ============================================

/**
 * Auto-generate tracking number if not provided
 */
shipmentSchema.pre('save', async function () {
    if (this.isNew && !this.trackingNumber) {
        this.trackingNumber = await generateTrackingNumber();
    }
});

// ============================================
// Instance Methods
// ============================================

/**
 * Add a new event to the timeline
 * @param {Object} event - Event object { status, location, description }
 * @returns {Promise<Shipment>}
 */
shipmentSchema.methods.addTimelineEvent = async function (event) {
    this.timeline.push({
        status: event.status,
        location: event.location || '',
        description: event.description || '',
        timestamp: event.timestamp || new Date(),
        isNotified: false,
    });

    // Update current status if event has a status
    if (event.updateStatus !== false) {
        this.status = event.status;
    }

    // Update current location if provided
    if (event.location) {
        this.currentLocation = this.currentLocation || {};
        this.currentLocation.city = event.location;
    }

    return this.save();
};

/**
 * Update shipment status with timeline entry
 * @param {string} newStatus - New status
 * @param {string} location - Location of status change
 * @param {string} description - Description of status change
 * @returns {Promise<Shipment>}
 */
shipmentSchema.methods.updateStatus = async function (newStatus, location = '', description = '') {
    return this.addTimelineEvent({
        status: newStatus,
        location,
        description,
    });
};

/**
 * Mark as picked up
 * @param {string} location - Pickup location
 * @returns {Promise<Shipment>}
 */
shipmentSchema.methods.markPickedUp = async function (location = '') {
    this.pickupDate = new Date();
    return this.updateStatus('Picked Up', location, 'Package picked up from sender');
};

/**
 * Mark as delivered
 * @param {string} recipientName - Name of recipient who signed
 * @returns {Promise<Shipment>}
 */
shipmentSchema.methods.markDelivered = async function (recipientName = '') {
    this.actualDelivery = new Date();
    if (recipientName) {
        this.recipientName = recipientName;
    }
    return this.updateStatus('Delivered', this.destination?.city || '', 'Package delivered successfully');
};

/**
 * Mark as exception
 * @param {string} reason - Exception reason
 * @param {string} location - Location of exception
 * @returns {Promise<Shipment>}
 */
shipmentSchema.methods.markException = async function (reason, location = '') {
    this.statusDetails = reason;
    return this.updateStatus('Exception', location, reason);
};

/**
 * Check if delivery is on time
 * @returns {boolean}
 */
shipmentSchema.methods.isOnTime = function () {
    if (!this.estimatedDelivery) return true;

    if (this.isDelivered) {
        return this.actualDelivery <= this.estimatedDelivery;
    }

    return new Date() <= this.estimatedDelivery;
};

// ============================================
// Static Methods
// ============================================

/**
 * Find by tracking number
 * @param {string} trackingNumber - Tracking number
 * @returns {Promise<Shipment|null>}
 */
shipmentSchema.statics.findByTracking = function (trackingNumber) {
    return this.findOne({ trackingNumber: trackingNumber.toUpperCase() })
        .populate('clientId', 'fullName email phone')
        .populate('requestId', 'requestNumber itemName');
};

/**
 * Find active shipments for a client
 * @param {ObjectId} clientId - Client ID
 * @returns {Promise<Shipment[]>}
 */
shipmentSchema.statics.findActiveByClient = function (clientId) {
    return this.find({
        clientId,
        status: { $nin: ['Delivered', 'Returned', 'Cancelled'] }
    }).sort({ createdAt: -1 });
};

/**
 * Find all shipments for a client
 * @param {ObjectId} clientId - Client ID
 * @returns {Promise<Shipment[]>}
 */
shipmentSchema.statics.findByClient = function (clientId) {
    return this.find({ clientId })
        .sort({ createdAt: -1 });
};

/**
 * Find shipments requiring attention (exceptions, delayed)
 * @returns {Promise<Shipment[]>}
 */
shipmentSchema.statics.findRequiringAttention = function () {
    const now = new Date();
    return this.find({
        $or: [
            { status: 'Exception' },
            {
                status: { $nin: ['Delivered', 'Returned', 'Cancelled'] },
                estimatedDelivery: { $lt: now }
            }
        ]
    })
        .populate('clientId', 'fullName email')
        .sort({ estimatedDelivery: 1 });
};

/**
 * Get shipment statistics for a client
 * @param {ObjectId} clientId - Client ID
 * @returns {Promise<Object>}
 */
shipmentSchema.statics.getClientStats = async function (clientId) {
    const stats = await this.aggregate([
        { $match: { clientId: new mongoose.Types.ObjectId(clientId) } },
        {
            $group: {
                _id: '$status',
                count: { $sum: 1 }
            }
        }
    ]);

    const result = {
        total: 0,
        active: 0,
        delivered: 0,
        inTransit: 0,
        exceptions: 0,
    };

    stats.forEach(s => {
        result.total += s.count;
        if (s._id === 'Delivered') {
            result.delivered = s.count;
        } else if (s._id === 'In Transit') {
            result.inTransit = s.count;
        } else if (s._id === 'Exception') {
            result.exceptions = s.count;
        } else if (!['Returned', 'Cancelled'].includes(s._id)) {
            result.active += s.count;
        }
    });

    return result;
};

// ============================================
// Helper Functions
// ============================================

/**
 * Generate unique tracking number
 * @returns {Promise<string>} - Generated tracking number (e.g., "LMS1234567890")
 */
async function generateTrackingNumber() {
    const prefix = 'LMS';
    let trackingNumber;
    let isUnique = false;

    while (!isUnique) {
        // Generate a 10-digit random number
        const randomPart = Math.floor(1000000000 + Math.random() * 9000000000).toString();
        trackingNumber = `${prefix}${randomPart}`;

        const existing = await mongoose.model('Shipment').findOne({ trackingNumber });
        if (!existing) {
            isUnique = true;
        }
    }

    return trackingNumber;
}

module.exports = mongoose.model('Shipment', shipmentSchema);
